[11.15] Added functionality to handle access tokens

inuitviking opened the pull request:
Closes #799 
Closes #798

This pull request includes:

- A new method in AbstractApi: getGroupPath (group equivalent to getProjectPath)
- New methods in Groups:
    - groupAccessTokens
    - groupAccessToken
    - createGroupAccessToken
    - deleteGroupAccessToken
    - rotateGroupAccessToken
- New methods in Projects
    - rotateProjectAccessToken
- Tests for above methods

The above methods are almost 1:1 to the same functionality available for project access tokens, making this pull request super easy to make. Tested with the available test tools seen in this project (albeit locally) and against my company's private group (with Premium). Works as expected in that regard.

The rotate methods simply rotate the specified access token, and returns the resulting array, containing also the token itself; this is handy for automatically updating variables in Gitlab containing these tokens.

inuitviking committed:
Added getGroupPath

This is similar to getProjectPath, but for groups.

inuitviking committed:
Added methods for group access tokens

Adds the same functions currently available for project access tokens,
but for groups.

inuitviking committed:
Added tests for group access tokens

inuitviking committed:
Fix PHPDoc style

m4tthumphrey commented:
@inuitviking Hello! Thanks for contributing on this project! However I am no longer involved with this, I think it is still under my username but I haven't used or look at this for a good few years now. I believe @GrahamCampbell is a maintainer these days.

inuitviking commented:
> @inuitviking Hello! Thanks for contributing on this project! However I am no longer involved with this, I think it is still under my username but I haven't used or look at this for a good few years now. I believe @GrahamCampbell is a maintainer these days.

Ah, sorry for bothering you! Thank you for mentioning Graham.

inuitviking committed:
Fixed PHP version requirements

The require for PHP versions was inaccurate, as this project actually
only supports PHP 8.1 and above.

inuitviking committed:
Added typed parameters and typed returns

This has been added to make PHPStan happi. :)

inuitviking committed:
Fix the doing of my IDE

It added tabs instead of spaces. :(

inuitviking committed:
Fixed PHPDocs in HttpClient/Pluging/*

PHPDocs were faulty (according to PHPStan) in Authentication and
ExceptionThrower. The `make test` command should no longer fail.

inuitviking committed:
Make StyleCI happy again.

It appears to be very picky; I guess this is good practice.

inuitviking committed:
Remove unsupported PHP versions from tests

Commit:
Add rotation functionality for group access tokens

Commit:
Made it possible to rotate project access tokens

GrahamCampbell commented:
Thanks for the PR. I've re-triggered the CI, and will try to review this shortly, for the 11.15.0 release. I wasn't able to allocate the time to get this into today's 11.14.0 release I'm afraid.

inuitviking commented:
I have added some more things since, as I got more requests for specific features. Will update tomorrow (Monday)

Commit:
Fixed styleci complaints

Commit:
Fix styleci complain that wasn't previously there

Commit:
Added tests for access token rotation

Commit:
Remove overseen comma

inuitviking commented:
> I have added some more things since, as I got more requests for specific features. Will update tomorrow (Monday)

By the way, and excuse my late comment, this has been done on that Monday.

cafferata commented on src/Api/Projects.php:
```diff
@@ -1715,6 +1715,25 @@ public function deleteProjectAccessToken($project_id, $token_id)
+     *
+     * @return mixed
+     */
+    public function rotateProjectAccessToken(string|int $project_id, string|int $token_id, string $expiry = ''): mixed
```
Crosslinking both (open) GitHub pull requests with the same method/functionality:
1. https://github.com/GitLabPHP/Client/pull/800/files#diff-160122af7affd3b85f6c02a3c1fcd7f20fbda356ab8bf3c89f71d0a56f68b31bR1725
2. https://github.com/GitLabPHP/Client/pull/808/files#diff-160122af7affd3b85f6c02a3c1fcd7f20fbda356ab8bf3c89f71d0a56f68b31bR1717

itrack commented:
Hi,
Can this pull request be merged?
Thank you!
